File: Trees/2583: Kth Largest Sum in a Binary Tree.py
# https://leetcode.com/problems/kth-largest-sum-in-a-binary-tree/
# difficulty: medium
# tags: bfs

# problem
# You are given the root of a binary tree and a positive integer k.

# The level sum in the tree is the sum of the values of the nodes that are on the same level.

# Return the kth largest level sum in the tree (not necessarily distinct). If there are fewer than k levels in the tree, return -1.

# Note that two nodes are on the same level if they have the same distance from the root.

# Solution, O(max(n, height log height)) time, O(n) space
# Get each level, sort, and return.

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def kthLargestLevelSum(self, root: Optional[TreeNode], k: int) -> int:
        queue = collections.deque()
        queue.append(root)
        levelSums = []
        while len(queue):
            queueLength = len(queue)
            levelSum = 0
            for i in range(queueLength):
                node = queue.popleft() # pretend O(1) queue
                levelSum += node.val
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            levelSums.append(levelSum)
        if k > len(levelSums):
            return -1
        levelSums.sort(reverse=True)
        return levelSums[k-1]

File: problems/to-categorize/Trees/2583: Kth Largest Sum in a Binary Tree.js
// https://leetcode.com/problems/kth-largest-sum-in-a-binary-tree/description/
// Difficulty: Medium
// tags: binary tree, bfs

// Problem
/*
You are given the root of a binary tree and a positive integer k.

The level sum in the tree is the sum of the values of the nodes that are on the same level.

Return the kth largest level sum in the tree (not necessarily distinct). If there are fewer than k levels in the tree, return -1.

Note that two nodes are on the same level if they have the same distance from the root.
*/

// Solution, O(n) time (with a real queue), O(n) space for the queue, O(levels log(levels)) time for the sort, O(levels) space for the sort, so the time can be O(n log n) for a linked list
// Just get the level sums, sort them, return the right one

var kthLargestLevelSum = function (root, k) {
  const levelSums = [];

  const queue = [root]; // fake queue

  while (queue.length > 0) {
    let sum = 0;
    const length = queue.length;
    for (let i = 0; i < length; i++) {
      const node = queue.shift();
      sum += node.val;
      if (node.left) {
        queue.push(node.left);
      }
      if (node.right) {
        queue.push(node.right);
      }
    }
    levelSums.push(sum);
  }

  levelSums.sort((a, b) => a - b);

  if (k > levelSums.length) {
    return -1;
  }

  return levelSums[levelSums.length - k];
};
